dartagnan/src/main/java/com/dat3m/dartagnan/verification/solving/RefinementSolver.java

```diff
@@ -605,13 +608,54 @@ private boolean process(Constraint c) {
+            empty (r & b) // b is now in a positive position and thus not need to get cut.
+
+        We claim that this model is equisatisfiable to the original memory model.
+        To understand this intuitively, it is helpful so think of the fresh b' as the original b but with "some slack".
+        We will argue that the best choice for b' is precisely b.
+        Any larger choice for b' will make r smaller, but since r is non-positive this reduces the chances
+        of satisfying negated axioms (e.g., if "~empty(r)" is true for some small r, then it is also true for all larger r).
+        On the other hand, if b' is chosen too small, then r gets so large that it violates "empty(r & b)".
+        We prove this formally:
+
+        "NEW => ORIGINAL":
+        Suppose M' is a consistent execution of the new memory model.
+        Then from "let r = a \ b'" it follows that M'(r) \subseteq M'(a).
+        Furthermore, from "empty (r & b)" it follows that M'(r) \subseteq ~M'(b).
+        Combining both, we get M'(r) \subseteq M'(a) & ~M'(b) = M'(a) \ M'(b)
+        Now consider an execution M of the original memory model that matches with M' on the base relations
+        (modulo b' which does not exist in the original model).
+        Then M(a) = M'(a) and M(b) = M'(b), so that M'(r) \subseteq M'(a) \ M'(b) = M(a) \ M(b) = M(r).
+        So the original memory model has a potentially larger interpretation for r (M'(r) \subseteq M(r)).
+        Since r is non-positive, we know that a larger relation satisfies more axioms (e.g.,
+        if r' is non-empty/non-acyclic/non-irreflexive, then the larger r will also satisfy these).
+        So M is a consistent execution of the original memory model.
+
+        "NEW <= ORIGINAL":
+        Suppose M is a consistent execution of the original model.
+        Then we can construct M' that matches with M on all common base relations but also sets M'(b') := M(b).
+        It is clear that M' now matches on all derived relations with M and so it satisfies the same axioms
+        and thus is also consistent.
+
+        NOTE: We argued about consistency only, but the same reasoning also applies to violation of flagged axioms,
+        i.e., the difference between "~empty(r)" and "flag ~empty(r)" is irrelevant for the argumentation.
+
+        NOTE 2: This rewriting is always possible.
+
+        NOTE 3: There is a minor caveat in the new memory model.
+        If an execution under the original memory model violates, e.g., "flag ~empty(r)" with multiple edges {e1, ..., ek},
+        then for the same execution the new memory model could report only a subset of those violations.
+        The reason is that the new memory model has a some slack in the choice of b':
+        it will choose b' such "flag ~empty(r)" holds (i.e., we have a violation), but it will not guarantee to
+        make the choice in such a way that it maximizes the number of violations.
```

Comment:
The SMT solver does not gives us this guarantee either, right?

Reply:
No, the text is meant to be seen from the SMT solver perspective. The SMT solver wants to find a satisfying assigment and so it will make sure that `~empty(r)` holds, but it will not care "how non-empty" `r` is.
That being said, I think the default search strategy of solvers prefer `false` over `true`, meaning the solver will automatically prefer to choose a small `b'` and thus a large `r`, maximizing the number of violations.

Btw., for a similar reason, if you have two flagged axioms, the solver might report the violation of only one of them even if both are violated in the original memory model.

Reply:
> Btw., for a similar reason, if you have two flagged axioms, the solver might report the violation of only one of them even if both are violated in the original memory model.

This is what I meant by "The SMT solver does not gives us this guarantee either"